gcs/constraints/smart_table.cc

```diff
@@ -0,0 +1,642 @@
+                    },
+                    [&](const UnaryValueEntry & unary_val_entry) {
+                        remove_supported(unsupported, unary_val_entry.var, supported_by_tree[unary_val_entry.var]);
+                    }}
+                    .visit(edge);
+            }
+        }
+    }
+
+    auto get_unrestricted(const vector<IntegerVariableID> & vars, const vector<SmartEntry> & tuple) -> vector<IntegerVariableID>
+    {
+        vector<IntegerVariableID> vars_in_tuple;
+        vector<IntegerVariableID> unrestricted;
+        for (const auto & entry : tuple) {
+            overloaded{
+                [&](const BinaryEntry & binary_entry) {
+                    vars_in_tuple.emplace_back(binary_entry.var_1);
+                },
+                [&](const UnarySetEntry & unary_set_entry) {
+                    vars_in_tuple.emplace_back(unary_set_entry.var);
+                },
+                [&](const UnaryValueEntry & unary_val_entry) {
+                    vars_in_tuple.emplace_back(unary_val_entry.var);
+                }}
+                .visit(entry);
+        }
+
+        auto vars_set = std::set(vars.begin(), vars.end());
+        auto vars_in_tuple_set = std::set(vars_in_tuple.begin(), vars_in_tuple.end());
+
+        std::set_difference(vars_set.begin(), vars_set.end(),
+            vars_in_tuple_set.begin(), vars_in_tuple_set.end(),
+            back_inserter(unrestricted));
+        return unrestricted;
+    }
+
+    auto propagate_using_smart_str(const vector<IntegerVariableID> & selectors, const vector<IntegerVariableID> & vars,
+        const SmartTuples & tuples, const vector<SmartTable::Forest> & forests, State & state) -> Inference
+    {
+
+        bool changed = false, contradiction = false;
+
+        VariableDomainMap unsupported{};
+        // Initialise unsupported values to everything in each variable's current domain.
+        for (const auto & var : vars) {
+            state.for_each_value(var, [&](Integer value) {
+                if (! unsupported.contains(var)) {
+                    unsupported[var] = vector<Integer>{};
+                }
+                unsupported[var].emplace_back(value);
+            });
+        }
+
+        // Check that feasible tuples are still feasible
+        // and also have them remove values from "unsupported" that they support
+        for (unsigned int tuple_idx = 0; tuple_idx < tuples.size(); ++tuple_idx) {
+            // Skip infeasible tuple
+            if (state.optional_single_value(selectors[tuple_idx]) == 0_i) {
+                continue;
+            }
+
+            for (const auto & tree : forests[tuple_idx]) {
+                // Initialise supported by tree to current variable domains
+                VariableDomainMap supported_by_tree{};
+
+                for (const auto & var : vars) {
+                    state.for_each_value(var, [&](Integer value) {
```

Comment:
Coming back to sorting: as it happens, `state.for_each_value` will always return them in sorted order. Perhaps we should make this an explicit guarantee? Maybe also we should have a `state.values_as_vector` which might be much faster?